keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1618 @@
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to invoke a local resource driver method. That method returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation.
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
+
+The flow is similar for a ResourceClaim that gets created as a stand-alone
+object by the user. In that case, the Pod reference that ResourceClaim by
+name. The ResourceClaim does not get deleted at the end and can be reused by
+another Pod and/or used by multiple different Pods at the same time (if
+supported by the driver). The resource remains allocated as long as the
+ResourceClaim doesn't get deleted.
+
+### API
+
+ResourceClaim  and ResourceClass are new built-in types
+in a new `cdi.k8s.io/v1alpha1` API group. This was chosen instead of
```

Comment:
The storage API group is called `storage.k8s.io` and not `csi.k8s.io`. Do you have alternative proposals? Could we re-use `node`?

Reply:
I had brought that up with @klueska. We agreed that reusing `node` would be feasible. I prefer a separate group because it feels cleaner and has some practical advantages (can have different OWNERS file with different reviewers, avoids code conflicts during development).

The name of course has to be usable. Another alternative would be `resource.k8s.io`. That is a bit more suitable, but also fairly generic.

Are there reasons (performance?) to avoid adding new groups?